ir/base.def

```diff
@@ -32,18 +32,25 @@ abstract Type {
     virtual const Type* getP4Type() const = 0;
 }
 
-/// Implemented by all types that may be generic
+/// Implemented by all types that may be generic:
+/// Control, Extern, Method, Package, and Parser
 interface IMayBeGenericType {
+    /// Allows the retrieval of type parameters
     virtual const TypeParameters* getTypeParameters() const = 0;
-    const Type* toType() const { return to<Type>(); }
+    // cast to a Type object.
+    // There is a single invocation of this method (typeChecker.cpp:92) so removed
+    // const Type* toType() const { return to<Type>(); }
 }
 
-/// Implemented by objects with an 'apply' method
+/// Implemented by objects with an 'apply' method: Parser, Control, Table
+/// \TODO: Unclear why the corresponding Type objects of these objects also implement IApply
```

Comment:
I assume that you mean Type_Parser and Type_Control.
I think that in the end only the Type_* objects have the required information for implementing this interface, P4Parser and P4Control only delegate the query to them. I think that you should think of Type_Parser as an interface, and as P4Parser as a class implementing the interface. Then IApply is just a base interface extended by Type_Parser.